pointmatcher/DataPointsFilters/utils/octree.h

```diff
@@ -0,0 +1,167 @@
+DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+*/
+#pragma once
+
+#include <cstdlib> 
+#include <vector>
+#include "PointMatcher.h"
+
+/*!
+ * \class octree.h
+ * \brief Octree class for DataPoints spatial representation
+ *
+ * \author Mathieu Labussiere (<mathieu dot labu at gmail dot com>)
+ * \date 24/05/2018
+ * \version 0.1
+ *
+ * Octree implementation for decomposing point cloud. 
+ * The current implementation use the data structure PointMatcher<T>::DataPoints. 
+ * It ensures that each node has either 8 or 0 childs. 
+ *
+ * Can create an octree with the 2 following crieterions:
+ *	- max number of data by node
+ *	- max size of a node (or stop when only one or zero data element is available)
+ *
+ * After construction, we can apply a process by creating a Visitor, implementing the following method:
+ *	```cpp
+ *	template<typename T>
+ *	struct Visitor {
+ *		bool operator()(Octree<T>& oc);
+ *	};
+ *	```
+ *
+ * Remark:
+ *	- Current implementation only store the indexes of the points from the pointcloud.
+ *	- Data element are exclusively contained in leaves node.
+ *	- Some leaves node contains no data (to ensure 8 or 0 childs).
+ *
+ */
+template < typename T >
+class Octree 
+{
+public:
+	using PM = PointMatcher<T>;
+	using DP = typename PM::DataPoints; /**/
+	using Id = typename DP::Index; /**/
+	
+	using Data = typename DP::Index; /**/
+	using DataContainer = std::vector<Data>;
+
+private:
+	struct XYZ 
+	{
+		T x;
+		T y;
+		T z;
+		
+		XYZ(T x_=T(0.0), T y_=T(0.0), T z_=T(0.0)) : x{x_}, y{y_}, z{z_} {}
+	
+		void operator+=(const XYZ& v){ x+=v.x; y+=v.y; z+=v.z; }
+		void operator-=(const XYZ& v){ x-=v.x; y-=v.y; z-=v.z; }
+		void operator*=(const XYZ& v){ x*=v.x; y*=v.y; z*=v.z; }
+		void operator*=(const T& s){ x*=s; y*=s; z*=s; }
+		void operator/=(const T& s){ x/=s; y/=s; z/=s; }
+		
+		XYZ operator+(const XYZ& v2){return XYZ{x+v2.x,y+v2.y,z+v2.z};}
+		XYZ operator-(const XYZ& v2){return XYZ{x-v2.x,y-v2.y,z-v2.z};}
+		XYZ operator*(const XYZ& v2){return XYZ{x*v2.x,y*v2.y,z*v2.z};}
+		XYZ operator*(const T& s)   {return XYZ{s*x,s*y,s*z};}
+		XYZ operator/(const XYZ& v2){return XYZ{x/v2.x,y/v2.y,z/v2.z};}
+		XYZ operator/(const T& s)   {return XYZ{s/x,s/y,s/z};}
+	};
+	
+	struct BoundingBox 
+	{
+			XYZ center;
+			T 	radius;
+	};
+	
+	Octree* parent;
+	Octree* octants[8];
+	
+	/******************************************************
+	 *	Octants id are assigned as their position 
+	 *   on a hamming cube (+ greater than center, - lower than center)
+	 *
+	 *	  	0	1	2	3	4	5	6	7
+	 * 	x:	-	+	-	+	-	+	-	+
+	 * 	z:	-	-	+	+	-	-	+	+
```

Comment:
Is there a special reason for this counterintuitive ordering? (z before y). If so maybe state it?

Reply:
In fact, I build the indexes as their position on a hamming cube, considering the axes as the same for 3D coordinates (direct system with the `000` point as origin). So the z coordinate is the 2nd bit in the index.

![hamming_cube](https://user-images.githubusercontent.com/38259866/40866539-2ace01c6-65cc-11e8-8c8c-e40258e98c8e.png)

If you think there is some drawback to use this notation, please tell me so I can change it.